Pocket.Logger/Tests/FormatterTests.cs

```diff
@@ -2,211 +2,225 @@
+              .Should()
+              .Be(" +[ (also, this) ]");
+    }
 
-            var result = formatter.Format(1, 2, 3, ("some-metric", 4));
+    [Fact]
+    public void When_there_are_additional_properties_then_they_are_appended_after_additional_args()
+    {
+        var formatter = Formatter.Parse("{one} and {two}");
 
-            result[2].Should().Be(("arg2", 3));
-            result[3].Should().Be(("arg3", ("some-metric", 4)));
-        }
+        var result = formatter.Format(
+            args: new object[] { 1, 2, 3, 4 },
+            knownProperties: new List<(string, object)>
+            {
+                ("also", "this")
+            });
 
-        [Fact]
-        public void IEnumerable_types_are_expanded()
-        {
-            var formatter = Formatter.Parse("The values in the array are: {values}");
+        result.ToString()
+              .Should()
+              .Be("1 and 2 +[ 3, 4, (also, this) ]");
+    }
 
-            var result = formatter.Format(new[] { 1, 2, 3, 4 }, new[] { 4, 5, 6 });
+    [Fact]
+    public void When_there_are_more_arguments_than_tokens_then_extra_arguments_are_available_as_key_value_pairs()
+    {
+        var formatter = Formatter.Parse("{one} and {two}");
 
-            result.ToString()
-                  .Should()
-                  .Be("The values in the array are: [ 1, 2, 3, 4 ] +[ [ 4, 5, 6 ] ]");
-        }
+        var result = formatter.Format(1, 2, 3, ("some-metric", 4));
 
-        [Fact]
-        public void Formatter_cache_stops_adding_items_after_specified_size()
-        {
-            Formatter.CacheLimit = 50;
+        result[2].Should().Be(("arg2", 3));
+        result[3].Should().Be(("arg3", ("some-metric", 4)));
+    }
 
-            for (var i = 0; i < 103; i++)
-            {
-                Formatter.Parse($"{i}");
-            }
+    [Fact]
+    public void IEnumerable_types_are_expanded()
+    {
+        var formatter = Formatter.Parse("The values in the array are: {values}");
+
+        var result = formatter.Format(new[] { 1, 2, 3, 4 }, new[] { 4, 5, 6 });
 
-            Formatter.CacheCount.Should().Be(50);
+        result.ToString()
+              .Should()
+              .Be("The values in the array are: [ 1, 2, 3, 4 ] +[ [ 4, 5, 6 ] ]");
+    }
+
+    [Fact]
+    public void Formatter_cache_stops_adding_items_after_specified_size()
+    {
+        Formatter.CacheLimit = 50;
+
+        for (var i = 0; i < 103; i++)
+        {
+            Formatter.Parse($"{i}");
         }
+
+        Formatter.CacheCount.Should().Be(50);
+    }
+
+    [Fact]
+    public void Formatter_can_be_customized_by_implementing_partial_method()
+    {
+        var entries = new List<string>();
+
+        using var subscription = LogEvents.Subscribe(
+            e => entries.Add(e.ToLogString()),
+            typeof(Class1).Assembly);
+
+        Class1.EmitSomeLogEvents(123);
+
+        entries.Should().Contain(e => e.Contains("parameter1 = 123 (custom)"));
```

Comment:
Consider: Add a test for the fallback i.e. where `CustomizeLogString` for some object explicitly returns `null` - this should fall back to calling `.ToString()`.

Also consider including a test for the case where `CustomizeLogString` returns `string.Empty` or whitespace. This should not invoke the fallback.